#27793 Adds list view mode to loader main UI

manneohrstrom opened the pull request:
This change adds a new "list view" mode to the main UI:

![image](https://cloud.githubusercontent.com/assets/337710/6653972/203717b4-ca9f-11e4-8f94-a3f7478575d8.png)

![image](https://cloud.githubusercontent.com/assets/337710/6653976/3658ac06-ca9f-11e4-993c-9c56eb88e564.png)

In addition to the above, there are some smaller tweaks as well:

- The delegate widgets have been moved into the loader so that it no longer depends on the standard widgets in the qtwidgets framework. This is a tweak to make it easier for the loader to switch to the v3 qtwidgets framework when that is released as part of the file manager changes.
- General cleanup around rendering of items
- Fixed an issue where actions were retriggered at exit.

jfboismenu-adsk commented on app.py:
```diff
@@ -18,6 +18,7 @@
 import sys
 import os
 
+
```
Typo or PEP8? :)

jfboismenu-adsk commented on python/tk_multi_loader/delegate_publish_list.py:
```diff
@@ -0,0 +1,343 @@
+
+class PublishListWidget(QtGui.QWidget):
+    """
+    Simple list *item* widget which hosts a square thumbnail, header text
```
Smells like a copy past from https://github.com/shotgunsoftware/tk-multi-loader2/blob/27793_list_view/python/tk_multi_loader/delegate_publish_history.py#L24

manneohrstrom commented:
Awesome! Thanks for the feedback! Let me make those tweaks and send some screenshots back for feedback. This was a quick tweak to at least being able to provide an alternative to the thumbnail view, but let me see if I can and get a search box added in at the same time. It would be super useful! :+1:

alandann commented on python/tk_multi_loader/delegate_publish_history.py:
```diff
@@ -17,6 +17,107 @@
 
+from .ui.widget_publish_history import Ui_PublishHistoryWidget
+
+class PublishHistoryWidget(QtGui.QWidget):
```
This should really be in it's own file

manneohrstrom replied:
This is how it started out - that was my gut feeling too - but it actually felt better once i had moved it inside this file - because the delegate and the widget are so related. I can move it out if you feel strongly about it, but in my opinion this would be a case where a good general rule of thumb is deliberately broken for improved readability! :)

alandann commented on python/tk_multi_loader/delegate_publish_history.py:
```diff
@@ -17,6 +17,107 @@
+        QtGui.QWidget.__init__(self, parent)
+
+        # make sure this widget isn't shown
+        self.setVisible(False)
```
I suspect I know why you are having to do this (I think I've just run into the same thing) - we should probably fix this in the base delegate class ;)

manneohrstrom replied:
This code is really old (i just moved it from the qtwidgets code as part of this pull request!) so I don't even fully remember why it's done this way. :) Do we have a list of shotgunutils/qtwidgets v3 stuff we should be doing? If not, let's huddle and get that list going - great opportunity here to fix some stuff in v2 which weren't ideal!

alandann commented on python/tk_multi_loader/delegate_publish_history.py:
```diff
@@ -17,6 +17,107 @@
+        self.ui.body_label.setText(body)
+
+    @staticmethod
+    def calculate_size():
```
Feels like this should be possible by overloading the correct method on the widget e.g. size, sizeHint, minimumSize, minimumSizeHint...  not exactly sure which one though ;)

manneohrstrom replied:
Another one for the v3 actions list!

alandann commented on python/tk_multi_loader/delegate_publish_history.py:
```diff
@@ -84,9 +204,7 @@ def _on_before_paint(self, widget, model_index, style_options):
-
-        if sg_item.get("version_number"):
-            header_str += "<b style='color:#2C93E2'>Version %03d</b>" % sg_item.get("version_number")
+        header_str += "<b style='color:#2C93E2'>Version %03d</b>" % (sg_item.get("version_number") or 0)
```
is sg_item a dictionary?  if so then sg_item.get("version_number", 0) would be better

manneohrstrom replied:
Because it is shotgun data, it is common that the dictionary data can be returned as None - this is why this `sg_data.get("field_name") or "default"` pattern is repeated everywhere. I didn't comment on it because it is something we do all the time. Does that make sense?

jfboismenu-adsk replied:
Could we put this on the technical debt list and see if we can do something for 1.0? This is so error prone than even a Toolkit veteran got it wrong. Not a good sign.

manneohrstrom replied:
I don't think this is technical debt – isn't it just code? A simple, pragmatic way to have it display "Version 0" instead of "Version None"?

jfboismenu-adsk replied:
Yes, but what I getting at is that I would have thought like Alan that sg_data.get("version_number", 0) would have worked. It's confusing to have None stored in the dictionary for that entry. That's my take on it anyway.

manneohrstrom replied:
It's the shotgun API basically - you ask it "find shots for me and return values name, in_frame and description" - and the data it returns is None for all those fields where no value has been set; 

```
{"type": "Shot", "id": 1234, "description": "Foo bar", "in_frame": None }
```

alandann commented on python/tk_multi_loader/delegate_publish_history.py:
```diff
@@ -95,18 +213,9 @@ def _on_before_paint(self, widget, model_index, style_options):
-            author_str = "%s" % sg_item.get("created_by").get("name")
-
+        desc_str = sg_item.get("description") or "No Description Given"         
+        author_str = sg_item.get("created_by").get("name") or "Unspecified User"
```
sg_item.get("created_by", {}).get("name") would be safer

alandann replied:
actually, I recently had a bug where a field from a shotgun query unexpectedly returned None instead of a dictionary - is it possible in this case that created_by may be None?  In which case splitting across multiple statements and checking each obtained value would be safer...

manneohrstrom replied:
I think `sg_item.get("created_by", {}).get("name")` would give a false sense of security, because it would be returned as `None` from shotgun if the value isn't defined but the field is requested! The code does make the assumption that `created_by` always exist - I think this is set for every record, whatever you do in Shotgun. I thought it was a safe assumption to make because created_by is such a special field, like creation_time, updated_by etc (I think they are called audit fields!). Does that make sense? Happy to put in additional checks, but only if you absolutely feel we need to!

jfboismenu-adsk replied:
If the intent is that creater_by should always be there, I think you should use [], otherwise it looks like you might have forgot to add a default value. Plus, if the create_by field is ever missing, the "KeyError: created_by" exception string is more telling than a "NoneType has no attribute get'.

manneohrstrom replied:
good point!

alandann commented on python/tk_multi_loader/delegate_publish_list.py:
```diff
@@ -0,0 +1,380 @@
+# Copyright (c) 2013 Shotgun Software Inc.
```
Very minor but copyright should be 2015

manneohrstrom replied:
will fix!

alandann commented on python/tk_multi_loader/delegate_publish_list.py:
```diff
@@ -0,0 +1,380 @@
+
+from .ui.widget_publish_list import Ui_PublishListWidget
+
+class PublishListWidget(QtGui.QWidget):
```
Should be in it's own file

manneohrstrom replied:
see above :)

alandann commented on python/tk_multi_loader/delegate_publish_list.py:
```diff
@@ -0,0 +1,380 @@
+        QtGui.QWidget.__init__(self, parent)
+
+        # make sure this widget isn't shown
+        self.setVisible(False)
```
Same comment as above ;)

manneohrstrom replied:
see above :)

alandann commented on python/tk_multi_loader/delegate_publish_list.py:
```diff
@@ -0,0 +1,380 @@
+        :param model_index: The model index to operate on
+        :param style_options: QT style options
+        """
+        # first set up the basic 
```
incomplete comment - doesn't make much sense!

manneohrstrom replied:
will fix!

alandann commented on python/tk_multi_loader/delegate_publish_list.py:
```diff
@@ -0,0 +1,380 @@
+            for v in field_value:
+                if isinstance(v, dict) and "name" in v and "type" in v:
+                    # This is a link field
+                    if v.get("name"):
```
this could be more readable and only do the dictionary look-up once, e.g.

name = v["name"]
if name:
    formatted_values.append(name)
    ...

manneohrstrom replied:
good point - will fix!

alandann commented on python/tk_multi_loader/delegate_publish_list.py:
```diff
@@ -0,0 +1,380 @@
+                else:
+                    formatted_values.append(str(v))
+            
+            if len(formatted_types) == 0:
```
This if/else isn't needed as types = ", ".join(formatted_types) works even if the set is empty (it will return an empty string)

manneohrstrom replied:
good point will fix! I think previously, there was something else going on in that if/else clause, which later on got removed, leaving us with this. Nice to clean up! :)

alandann commented on python/tk_multi_loader/delegate_publish_list.py:
```diff
@@ -0,0 +1,380 @@
+        created_unixtime = sg_data.get("created_at") or 0
+        date_str = datetime.datetime.fromtimestamp(created_unixtime).strftime('%Y-%m-%d %H:%M')
+        small_text = "Version %03d by %s at %s" % (sg_data.get("version_number"), 
+                                                      sg_data.get("created_by").get("name"),
```
same here - are you sure "created_by" can't be None?

manneohrstrom replied:
see above! :)

jfboismenu-adsk replied:
See above for my comment too! :)

alandann commented on python/tk_multi_loader/delegate_publish_thumb.py:
```diff
@@ -16,40 +17,176 @@
 
-class SgPublishDelegate(shotgun_view.WidgetDelegate):
+
+class PublishThumbWidget(QtGui.QWidget):
```
Should be in it's own file

manneohrstrom replied:
see above! :)

alandann commented on python/tk_multi_loader/delegate_publish_thumb.py:
```diff
@@ -16,40 +17,176 @@
+        
+        # set up an event filter to ensure that the thumbnails
+        # are scaled in a square fashion.
+        filter = ResizeEventFilter(self.ui.thumbnail)
```
ResizeEventFilter is a really ambiguous name - maybe make it more specific or move it inside this class as a protected sub-class?

manneohrstrom replied:
I thought it was enough to keep it in the same file to indicate that this class was internal to the .py file and wasn't meant to be used anywhere else! I don't generally like nested classes but it's a good point - i'll shift it round :)

alandann commented on python/tk_multi_loader/delegate_publish_thumb.py:
```diff
@@ -165,26 +368,49 @@ def _on_before_paint(self, widget, model_index, style_options):
+# utility classes
+
+
+class ResizeEventFilter(QtCore.QObject):
```
see comment above where this is used

alandann commented on python/tk_multi_loader/delegate_publish_thumb.py:
```diff
@@ -165,26 +368,49 @@ def _on_before_paint(self, widget, model_index, style_options):
+    """
+    resized = QtCore.Signal()
+
+    def eventFilter(self, obj, event):
```
Missing doc-string

manneohrstrom replied:
will fix

alandann commented on python/tk_multi_loader/dialog.py:
```diff
@@ -363,6 +379,46 @@ def _on_history_double_clicked(self, model_index):
+        :param mode: either MAIN_VIEW_LIST or MAIN_VIEW_THUMB
+        """
+        if mode == self.MAIN_VIEW_LIST:
+            self.ui.list_mode.setIcon(QtGui.QIcon(QtGui.QPixmap(":/res/mode_switch_card_active.png")))
```
not major but this could be done through the style sheet instead (by setting the icons on the sub-states)

alandann commented on python/tk_multi_loader/dialog.py:
```diff
@@ -363,6 +379,46 @@ def _on_history_double_clicked(self, model_index):
+            self.ui.list_mode.setIcon(QtGui.QIcon(QtGui.QPixmap(":/res/mode_switch_card_active.png")))
+            self.ui.list_mode.setChecked(True)
+            self.ui.thumbnail_mode.setIcon(QtGui.QIcon(QtGui.QPixmap(":/res/mode_switch_thumb.png")))
+            self.ui.thumbnail_mode.setChecked(False)
```
Using radio buttons would also reduce the code here

manneohrstrom replied:
can you make radio buttons look the same? The current setup is pretty clunky, i'll see if i can simplify it with some style sheeting and will try the radio buttons...

manneohrstrom replied:
here's a recipe (thanks @alandann)

```
#grid_radio_btn::indicator {
width: 26;
height: 24;
}

#grid_radio_btn::indicator::unchecked {
image: url(:/tk-multi-workfiles/grid_view_unchecked.png);
}

#grid_radio_btn::indicator::unchecked::hover {
image: url(:/tk-multi-workfiles/grid_view_unchecked_hover.png);
}

#grid_radio_btn::indicator::checked {
    image: url(:/tk-multi-workfiles/grid_view_checked.png);
}
```

alandann commented on python/tk_multi_loader/dialog.py:
```diff
@@ -363,6 +379,46 @@ def _on_history_double_clicked(self, model_index):
+        else:
+            raise TankError("Undefined view mode!") 
+
+        self.ui.publish_view.selectionModel().clear()
```
the behaviour that switching views clears the selection seems odd to me

manneohrstrom replied:
It's a delegate weirdness thing. i'll add a comment and will add it to the v3 doc for the libraries :)

alandann commented:
cr done - a few comments but nothing major

manneohrstrom commented:
Great comments @alandann will make the changes - left some of them as questions to check what you think. Will also make the adjustments that rodeo suggested for the formatting and report back!

manneohrstrom commented:
Addressed Code review and tweaked the UI based on @rodeo 's suggestions:

![image](https://cloud.githubusercontent.com/assets/337710/7090771/5a0c82ae-df9d-11e4-9c06-c8caf2246bb9.png)
